### PR TITLE
Updating sites module

### PR DESCRIPTION
**Pull request by dr-dolittle**
# Change

Changes:
- Adding comments
- Updating readmes
- Adding linux kind
- Updating app settings
- Adding auth settings

Regarding auth settings: there are a couple of topics / issues that need to be resolved before this module can be fully tested:

1. The Pester test fails due to `<<tenantId>>` (even when parameter `authSettingV2Configuration` is commented in the parameters file).
2. To be able to test easy auth it is necessary to have an app id (not SP object id), and this app needs to have an API exposed (e.g. `api://<<app id>>`) as well as a secret included, that is stored in a Key Vault to be referenced via the app settings.

How do we proceed here?

[Web: Sites](https://github.com/quervernetzt/ResourceModules/actions/workflows/ms.web.sites.yml)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

**Comment by dr-dolittle** [comment]
@MrMCake here the pull request. Please check. Thanks

File: arm/Microsoft.Web/sites/config-appsettings/deploy.bicep [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+@description('Optional. The app settings key value pairs except for AzureWebJobsStorage, AzureWebJobsDashboard, APPINSIGHTS_INSTRUMENTATIONKEY and APPLICATIONINSIGHTS_CONNECTION_STRING.')
+param appSettingsKeyValuePairs object = {}
+
+@description('Optional. Enable telemetry via the Customer Usage Attribution ID (GUID).')
+param enableDefaultTelemetry bool = true
+
+// =========== //
+// Variables   //
+// =========== //
+var azureWebJobsValues = !empty(storageAccountId) && setAzureWebJobsDashboard == true ? {
```
Comment (MrMCake) [inline arm/Microsoft.Web/sites/config-appsettings/deploy.bicep:33]: Isn't it easier like this?

```bicep
var azureWebJobsValues = !empty(storageAccountId) ? union({
  AzureWebJobsStorage: 'DefaultEndpointsProtocol=https;AccountName=${storageAccount.name};AccountKey=${storageAccount.listKeys().keys[0].value};'
}, ((setAzureWebJobsDashboard == false) ? {
  AzureWebJobsDashboard: 'DefaultEndpointsProtocol=https;AccountName=${storageAccount.name};AccountKey=${storageAccount.listKeys().keys[0].value};'
} : {})) : {} 
```

Comment (dr-dolittle) [reply arm/Microsoft.Web/sites/config-appsettings/deploy.bicep]: @MrMCake Thanks for the suggestion. I updated it.

**Comment by MrMCake** [comment]
@MariusStorhaug / @eriqua any idea regarding the AuthSettings (see description).
We now need an ApplicationID & AppSecret. So the MSI/deploymentSpId won't help us

**Comment by dr-dolittle** [comment]
> @MariusStorhaug / @eriqua any idea regarding the AuthSettings (see description). We now need an ApplicationID & AppSecret. So the MSI/deploymentSpId won't help us

+ API exposed

@MariusStorhaug / @eriqua / @MrMCake 
How do we proceed here?

Thanks

**Comment by eriqua** [comment]
> > @MariusStorhaug / @eriqua any idea regarding the AuthSettings (see description). We now need an ApplicationID & AppSecret. So the MSI/deploymentSpId won't help us
> 
> * API exposed
> 
> @MariusStorhaug / @eriqua / @MrMCake How do we proceed here?
> 
> Thanks

@MrMCake / @dr-dolittle Could you please elaborate a bit more on this? What is the question exactly?

**Comment by MrMCake** [comment]
> @MariusStorhaug / @eriqua any idea regarding the AuthSettings (see description). We now need an ApplicationID & AppSecret. So the MSI/deploymentSpId won't help us

> @MrMCake / @dr-dolittle Could you please elaborate a bit more on this? What is the question exactly?

The issue is to be found in the modules parameter file:

image

We have cases where we need both a ServicePrincipal _**Application ID**_ and a secret. However, both are not available in our environment as test values / tokens.

**Comment by eriqua** [comment]
> > @MariusStorhaug / @eriqua any idea regarding the AuthSettings (see description). We now need an ApplicationID & AppSecret. So the MSI/deploymentSpId won't help us
> 
> > @MrMCake / @dr-dolittle Could you please elaborate a bit more on this? What is the question exactly?
> 
> The issue is to be found in the modules parameter file:
> 
> image
> 
> We have cases where we need both a ServicePrincipal _**Application ID**_ and a secret. However, both are not available in our environment as test values / tokens.

So would it work to add its creation to the dependency pipeline, backup those required parameters as secrets in our shared kv and reference those in the parameter file?

**Comment by MariusStorhaug** [comment]
We do, the deployment son, and the secret that is stored in the AZURE_CREDENTIALS.CLIENT_SECRET + CLIENT_ID, but guess we could also just make another to test with.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by MariusStorhaug** [comment]
Somone please start a thread :D 
@eriqua I think it would work to create an SPN get its secret and store it. It requires that we also add the permissions it needs i AAD to do so in that case. On the other hand, cant we use the info that is stored in the AZURE_CREDENTIALS? There we have CLIENT_SECRET + CLIENT_ID.

**Comment by dr-dolittle** [comment]
> Someone please start a thread :D @eriqua I think it would work to create an SPN get its secret and store it. It requires that we also add the permissions it needs i AAD to do so in that case. On the other hand, cant we use the info that is stored in the AZURE_CREDENTIALS? There we have CLIENT_SECRET + CLIENT_ID.

How do I start a thread here?

What I need is
- TenantId -> I can get this via <<tenantId>> -> but one of the Pester tests fails -> `Expected $true, because Parameter file should not contain the Tenant ID guid, instead should reference a token value '<<tenantId>>', but got $false.` but I use `<<tenantId>>` -> needs to be fixed.
- An AAD app with client id and client secret (stored in the KeyVault) as well as an API endpoint enabled (e.g. api://<<app id>>). Once I have the client id, the key vault reference to the secret and the API endpoint I can test the deployment.

Please let me know once these things available and how I can access the data.

Thanks!

**Comment by MrMCake** [comment]
> Someone please start a thread :D @eriqua I think it would work to create an SPN get its secret and store it. It requires that we also add the permissions it needs i AAD to do so in that case. On the other hand, cant we use the info that is stored in the AZURE_CREDENTIALS? There we have CLIENT_SECRET + CLIENT_ID.

This won't work for Azure DevOps unfortunately

**Comment by MrMCake** [comment]
> > Someone please start a thread :D @eriqua I think it would work to create an SPN get its secret and store it. It requires that we also add the permissions it needs i AAD to do so in that case. On the other hand, cant we use the info that is stored in the AZURE_CREDENTIALS? There we have CLIENT_SECRET + CLIENT_ID.
> 
> @MariusStorhaug
> 
> How do I start a thread here?
> 
> What I need is
> 
> * TenantId -> I can get this via `<<tenantId>>` -> but one of the Pester tests fails -> `Expected $true, because Parameter file should not contain the Tenant ID guid, instead should reference a token value '<<tenantId>>', but got $false.` but I use `<<tenantId>>` -> needs to be fixed.
> * An AAD app with client id and client secret (stored in the KeyVault) as well as an API endpoint enabled (e.g. `api://<<app id>>`). Once I have the client id, the key vault reference to the secret and the API endpoint I can test the deployment.
> 
> Please let me know once these things available and how I can access the data.
> 
> Thanks!

The issue with the tenantID should be fixed in main. Please merge from upstream

**Comment by dr-dolittle** [comment]
> > > Someone please start a thread :D @eriqua I think it would work to create an SPN get its secret and store it. It requires that we also add the permissions it needs i AAD to do so in that case. On the other hand, cant we use the info that is stored in the AZURE_CREDENTIALS? There we have CLIENT_SECRET + CLIENT_ID.
> > 
> > 
> > @MariusStorhaug
> > How do I start a thread here?
> > What I need is
> > 
> > * TenantId -> I can get this via `<<tenantId>>` -> but one of the Pester tests fails -> `Expected $true, because Parameter file should not contain the Tenant ID guid, instead should reference a token value '<<tenantId>>', but got $false.` but I use `<<tenantId>>` -> needs to be fixed.
> > * An AAD app with client id and client secret (stored in the KeyVault) as well as an API endpoint enabled (e.g. `api://<<app id>>`). Once I have the client id, the key vault reference to the secret and the API endpoint I can test the deployment.
> > 
> > Please let me know once these things available and how I can access the data.
> > Thanks!
> 
> The issue with the tenantID should be fixed in main. Please merge from upstream

@MrMCake I merged from upstream but fails -> https://github.com/quervernetzt/ResourceModules/actions/runs/2189803727

image

**Comment by dr-dolittle** [comment]
> > https://github.com/quervernetzt/ResourceModules
> 
> @dr-dolittle is this still the case? Just checked out the repository & branch and all 110 tests passed:
> 
> > Tests completed in 39.9s
> > Tests Passed: 110, Failed: 0, Skipped: 0 NotRun: 9

@MrMCake I enabled the Actions again. Check out https://github.com/quervernetzt/ResourceModules/actions/runs/2189803727. Similar to the AML module this one fails.

**Comment by MrMCake** [comment]
@dr-dolittle I also had a quick chat with @eriqua. We'd suggest to create a SPN manually (I can do that today) and push its credentials into the dependency key vault. That said, this requires 2 things
1. It should be documented in a subsection of the [Getting Started Guide](https://github.com/Azure/ResourceModules/wiki/Getting%20started%20-%20Scenario%201%20Onboard%20module%20library%20and%20CI%20environment#4-deploy-dependencies). I guess something like `4.1 Manual dependencies`. We'd need to explain what needs to be done and for which modules this needs to be done (alongside a disclaimer that says : Beware that the SPN secret expires after x amount of months)
2. JSON parameter files do not support secret references in a nested way (i.e. you can only use a reference as a child of the `"parameters": {}` object in the `parameters.json` file, not in a nested parameter of such a parameter. That means, in order to make this work we'd either need to change to implementation and ask the user for the secret-resourceID and fetch it in the bicep file using Bicep native GetSecret() functions, or move the reference out of the AuthV2 parameter to the top level.

File: arm/Microsoft.Web/sites/deploy.bicep [hunk 2]
```diff
@@ -181,24 +186,37 @@ resource app 'Microsoft.Web/sites@2020-12-01' = {
...
   params: {
-    name: 'appsettings'
     appName: app.name
-    storageAccountId: !empty(storageAccountId) ? storageAccountId : ''
-    appInsightId: !empty(appInsightId) ? appInsightId : !empty(appInsightObject) ? appInsight.outputs.resourceId : ''
-    functionsWorkerRuntime: !empty(functionsWorkerRuntime) ? functionsWorkerRuntime : ''
-    functionsExtensionVersion: !empty(functionsExtensionVersion) ? functionsExtensionVersion : '~3'
+    kind: kind
+    authSettingV2Configuration: authSettingV2Configuration
+    enableDefaultTelemetry: enableDefaultTelemetry
```
Comment (MrMCake) [inline arm/Microsoft.Web/sites/deploy.bicep:219]:
```suggestion
```

**Comment by dr-dolittle** [comment]
@MrMCake @eriqua @MariusStorhaug I implemented the changes (including the REAMDE) and the pipelines on their own are completing successfully. But there are currently two issues:

1. The Linter tests fail -> can you please check what to do about it (https://github.com/Azure/ResourceModules/runs/6117665848?check_suite_focus=true)?
2. We have 4 parameter files, i.e. 4 deployments that use the same App Service Plan. Unfortunately, not all deployments run at the same time and as you can see here (https://github.com/quervernetzt/ResourceModules/runs/6118224952?check_suite_focus=true), the App Service Plan is being deleted as part of a previous Function App deletion and is therefore not available for further deployments (of the same pipeline run). We would need e.g. a function app that is deployed as dependency using the ASP so that there is always an app connected to the ASP. What do you think about this approach? How would you resolve this issue?

Thanks

**Comment by dr-dolittle** [comment]
@MrMCake @eriqua @MariusStorhaug The pipeline run was now successful (not removing the resources) -> https://github.com/quervernetzt/ResourceModules/actions/runs/2206131048. Can we now merge? Thanks

File: docs/wiki/Getting started - Scenario 1 Onboard module library and CI environment.md [hunk 3]
```diff
@@ -330,6 +330,14 @@ Run the dependencies pipeline by following instructions provided in the specific
 
 > **Note**: For details about the dependencies pipeline design please refer to the dedicated [Dependencies pipeline design](./The%20CI%20environment%20-%20Pipeline%20design.md#dependencies-pipeline) section.
 
+## 4.1 Manual Dependencies
+
+
+### Microsoft.Web/sites
+
+To successfully deploy the function app using the parameter file `fa.parameters.json` you need to create an Azure Active Directory App with API endpoint enabled (e.g. `api://<app id>`) and a secret added. The secret value needs then to be stored in a Key Vault secret.
```
Comment (MrMCake) [inline docs/wiki/Getting started - Scenario 1 Onboard module library and CI environment.md:338]:
```suggestion
To successfully deploy the sites module using the parameter file `fa.parameters.json` you need to create an Azure Active Directory App with its API endpoint enabled (e.g. `api://<app id>`) and add a secret. The secret value needs then to be stored in a Key Vault secret.
```

File: docs/wiki/Getting started - Scenario 1 Onboard module library and CI environment.md [hunk 4]
```diff
@@ -330,6 +330,14 @@ Run the dependencies pipeline by following instructions provided in the specific
 
 > **Note**: For details about the dependencies pipeline design please refer to the dedicated [Dependencies pipeline design](./The%20CI%20environment%20-%20Pipeline%20design.md#dependencies-pipeline) section.
 
+## 4.1 Manual Dependencies
```
Comment (MrMCake) [inline docs/wiki/Getting started - Scenario 1 Onboard module library and CI environment.md:333]:
```suggestion
## 4.1 Manual Dependencies

In special cases, manual actions may be required to provision certain resources that are not covered by our dependency pipeline. In the following you can find an overview of these resources, for which modules you'd need them and what you'd need to do:
```

File: docs/wiki/Getting started - Scenario 1 Onboard module library and CI environment.md [hunk 5]
```diff
@@ -330,6 +330,14 @@ Run the dependencies pipeline by following instructions provided in the specific
...
+
+
```
Comment (MrMCake) [inline docs/wiki/Getting started - Scenario 1 Onboard module library and CI environment.md:335]:
```suggestion
```

**Review by MrMCake — APPROVED** [review]

**Pull request merged**